File: packages/bear_necessities/lib/app/modules/shared/utils/card_id_obfuscator.dart
/// Deterministic card-ID obfuscation shared between the Flutter app and the
/// QR / Excel generator scripts.
///
/// MUST stay byte-for-byte identical to
/// ``scripts/card_id_obfuscator.py``. See that file for the rationale.
///
/// Properties of the encoding:
/// * Deterministic, bijective 1:1 mapping between plaintext card IDs
///   (``TODO_GG_07``, ``TODO_GGX_15``, ...) and 4-character codes such as
///   ``AZJJ`` / ``BXHE``.
/// * Strong avalanche — neighbouring IDs produce completely unrelated codes.
/// * Codes use an unambiguous base32 alphabet (no ``0/O``, ``1/I``) and are
///   URL-safe.
///
/// Decoding is case-insensitive so deep links remain robust against
/// well-meaning URL normalisation.
library;

class CardIdObfuscator {
  CardIdObfuscator._();

  // --------------------------------------------------------------------
  // Tunables — MUST match scripts/card_id_obfuscator.py byte-for-byte.
  // --------------------------------------------------------------------

  /// Four 40-bit round keys for the Feistel cipher. Changing any of these
  /// invalidates every QR code currently in the wild — DO NOT change.
  static const List<int> _masterKey = <int>[
    0xA5C3F19E72,
    0x3F1C5A92B4,
    0x9E724B1F8C,
    0xC73E5A91F4,
  ];

  /// 32-symbol base32 alphabet without visually ambiguous characters
  /// (no ``0/O``, no ``1/I``).
  static const String _alphabet = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";

  /// Fixed code length in base32 chars. 16 chars × 5 bits/char = 80 bits,
  /// exactly the Feistel block size so every cipher output fits with no
  /// leading padding.
  static const int _codeLen = 16;

  // Feistel block geometry. 40-bit halves with all shift counts <= 13 keep
  // every pre-mask intermediate below 2**53, which is Flutter Web's safe
  // integer ceiling.
  static const int _halfBits = 40;
  static const int _halfMask = (1 << _halfBits) - 1; // 0xFFFFFFFFFF

  // ID-space partitioning. Keep in sync with the Python module.
  static const String _prefixGg = "TODO_GG_";
  static const String _prefixGgx = "TODO_GGX_";
  static const int _ggxOffset = 100;

  // --------------------------------------------------------------------
  // Public API
  // --------------------------------------------------------------------

  /// Encode a plaintext card ID into its obfuscated 16-char code.
  ///
  /// Throws [ArgumentError] for unknown card IDs.
  static String encode(String cardId) {
    final int n = _idToInt(cardId);
    // The plaintext lives in the low 40 bits; the high half is zero.
    final List<int> halves = _encrypt((n >> _halfBits) & _halfMask, n & _halfMask);
    return _halfToBase32(halves[0]) + _halfToBase32(halves[1]);
  }

  /// Decode an obfuscated code back into its plaintext card ID.
  ///
  /// Returns ``null`` for malformed codes or codes that do not correspond
  /// to a known card. Case-insensitive.
  static String? tryDecode(String code) {
    if (code.length != _codeLen) return null;
    final int? leftCipher = _halfFromBase32(code.substring(0, _halfCodeLen));
    final int? rightCipher = _halfFromBase32(code.substring(_halfCodeLen));
    if (leftCipher == null || rightCipher == null) return null;
    final List<int> plain = _decrypt(leftCipher, rightCipher);
    // Slot must fit in 40 bits — anything in the high half means bogus input.
    if (plain[0] != 0) return null;
    return _tryIntToId(plain[1]);
  }

  /// Whether [value] looks like an obfuscated code (right length and only
  /// alphabet chars). Does not verify that it decodes to a known card.
  static bool looksLikeCode(String value) {
    if (value.length != _codeLen) return false;
    final upper = value.toUpperCase();
    for (int i = 0; i < upper.length; i++) {
      if (!_alphabet.contains(upper[i])) return false;
    }
    return true;
  }

  // --------------------------------------------------------------------
  // 80-bit Feistel cipher (two 40-bit halves)
  //
  // Uses only XOR / add / shift / rotate so the Python port mirrors it
  // verbatim. Every shift count is <=13 so the worst-case pre-mask
  // intermediate is ``(2^40 - 1) * 2^13 < 2^53`` — safe on Flutter Web's
  // JS-number integers.
  // --------------------------------------------------------------------

  /// Rotate a 40-bit value left by [n] bits. The mask + cap on [n] guarantee
  /// the result fits in 40 bits and the pre-mask intermediate fits in 53.
  static int _rotl40(int x, int n) {
    n %= _halfBits;
    if (n == 0) return x & _halfMask;
    final int left = (x << n) & _halfMask;
    final int right = x >> (_halfBits - n);
    return (left | right) & _halfMask;
  }

  static int _f(int half, int k) {
    int x = (half + k) & _halfMask;
    x ^= _rotl40(x, 13);
    x = (x + ((x << 7) & _halfMask)) & _halfMask;
    x ^= _rotl40(x, 11);
    x = (x + ((x << 5) & _halfMask)) & _halfMask;
    x ^= _rotl40(x, 9);
    return x & _halfMask;
  }

  /// Forward Feistel network. Inputs and outputs are two 40-bit halves so
  /// the result fits in Dart's 64-bit int (the combined 80-bit value would
  /// not). Returned as ``[left, right]``.
  static List<int> _encrypt(int left, int right) {
    left &= _halfMask;
    right &= _halfMask;
    for (final k in _masterKey) {
      final int newRight = (left ^ _f(right, k)) & _halfMask;
      left = right;
      right = newRight;
    }
    return <int>[left, right];
  }

  /// Inverse of [_encrypt]. Returns ``[left, right]``.
  static List<int> _decrypt(int left, int right) {
    left &= _halfMask;
    right &= _halfMask;
    for (int i = _masterKey.length - 1; i >= 0; i--) {
      final int k = _masterKey[i];
      // Undo (L,R) -> (R, L XOR F(R,k))
      final int prevRight = left;
      final int prevLeft = (right ^ _f(left, k)) & _halfMask;
      left = prevLeft;
      right = prevRight;
    }
    return <int>[left, right];
  }

  // --------------------------------------------------------------------
  // Base32 helpers
  //
  // Each 40-bit Feistel half encodes to exactly 8 base32 chars; the two
  // halves are concatenated (high half first) to form the 16-char code.
  // --------------------------------------------------------------------

  static const int _halfCodeLen = _halfBits ~/ 5; // = 8

  /// Encode a 40-bit unsigned int as exactly 8 base32 chars (big-endian).
  static String _halfToBase32(int n) {
    assert(n >= 0 && n <= _halfMask);
    final chars = List<String>.filled(_halfCodeLen, "");
    for (int i = _halfCodeLen - 1; i >= 0; i--) {
      chars[i] = _alphabet[n & 31];
      n >>= 5;
    }
    return chars.join();
  }

  /// Decode an 8-char base32 chunk into a 40-bit int, or ``null`` if any
  /// character is invalid. Case-insensitive.
  static int? _halfFromBase32(String code) {
    int n = 0;
    final upper = code.toUpperCase();
    for (int i = 0; i < upper.length; i++) {
      final int idx = _alphabet.indexOf(upper[i]);
      if (idx < 0) return null;
      n = (n << 5) | idx;
    }
    return n & _halfMask;
  }

  // --------------------------------------------------------------------
  // ID <-> integer mapping
  // --------------------------------------------------------------------

  static int _idToInt(String cardId) {
    if (cardId.startsWith(_prefixGgx)) {
      final suffix = cardId.substring(_prefixGgx.length);
      final idx = int.tryParse(suffix);
      if (idx == null || idx < 1 || idx > 30) {
        throw ArgumentError.value(cardId, "cardId", "GGX index out of range");
      }
      return _ggxOffset + idx;
    }
    if (cardId.startsWith(_prefixGg)) {
      final suffix = cardId.substring(_prefixGg.length);
      final idx = int.tryParse(suffix);
      if (idx == null || idx < 1 || idx > 60) {
        throw ArgumentError.value(cardId, "cardId", "GG index out of range");
      }
      return idx;
    }
    throw ArgumentError.value(cardId, "cardId", "unknown card-ID shape");
  }

  static String? _tryIntToId(int n) {
    if (n >= 1 && n <= 60) {
      return "$_prefixGg${n.toString().padLeft(2, '0')}";
    }
    if (n >= _ggxOffset + 1 && n <= _ggxOffset + 30) {
      final idx = n - _ggxOffset;
      return "$_prefixGgx${idx.toString().padLeft(2, '0')}";
    }
    return null;
  }
}

File: scripts/card_id_obfuscator.py
"""Deterministic card-ID obfuscation shared between scripts and the app.

The Bear Adventure app accepts deep links of the form
``https://app.bearfruitsnacks.com/?path=<code>`` to unlock a card. The
plaintext card IDs (``TODO_GG_07``, ``TODO_GGX_15``, ...) follow a trivially
guessable pattern, so we obfuscate them into 16-character codes such as
``7NMDSYKYSH4TNEYF`` before printing them on a QR.

Properties of the encoding:

* **Deterministic & bijective.** Each ``card_id`` maps to exactly one
  ``code`` and back. Encoding the same ID twice always produces the same code.
* **Avalanche.** Consecutive IDs (``TODO_GG_07`` vs ``TODO_GG_08``) produce
  wildly different codes; you cannot infer one code from a neighbouring one.
* **Compact.** Codes are always 16 characters from an unambiguous base32
  alphabet (no ``0/O``, ``1/I``, etc.), making them URL-safe and easy to
  inspect manually.
* **Portable.** The algorithm is a 4-round Feistel cipher on an 80-bit block
  (two 40-bit halves) using only XOR / add / shift / rotate, so it ports
  cleanly to Flutter Web's 53-bit JS-number integers. The Dart port lives in
  ``packages/bear_necessities/lib/app/modules/shared/utils/card_id_obfuscator.dart``
  and MUST stay byte-for-byte identical to this file.

If you ever need to introduce new card-ID shapes, extend ``_id_to_int`` /
``_int_to_id`` in BOTH this file and the Dart port simultaneously, and add a
roundtrip test for the new shape.
"""

from __future__ import annotations

# ---------------------------------------------------------------------------
# Tunables — MUST match the Dart port byte-for-byte.
# ---------------------------------------------------------------------------

# Four 40-bit round keys for the Feistel cipher. Changing any of these
# invalidates every QR code currently in the wild — DO NOT change.
_MASTER_KEY: tuple[int, int, int, int] = (
    0xA5C3F19E72,
    0x3F1C5A92B4,
    0x9E724B1F8C,
    0xC73E5A91F4,
)

# 32-symbol base32 alphabet without visually ambiguous characters
# (no 0/O, no 1/I). Decoders accept lower-case input as well.
_ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
assert len(_ALPHABET) == 32

_INDEX: dict[str, int] = {c: i for i, c in enumerate(_ALPHABET)}

# How many base32 chars per code. 16 chars * 5 bits/char = 80 bits, which is
# exactly the Feistel block size so every cipher output fits with no leading
# padding (avoiding a recognizable "always-A" prefix).
_CODE_LEN = 16

# Feistel block geometry. 40-bit halves keep every intermediate value safely
# below 2**53 so the Dart port stays correct on Flutter Web's JS-number
# integers. The round function therefore avoids large multiplications and
# uses only XOR / add / shift / rotate.
_HALF_BITS = 40
_HALF_MASK = (1 << _HALF_BITS) - 1
_BLOCK_BITS = _HALF_BITS * 2

# ID-space partitioning. Each known card-ID prefix is mapped to a contiguous
# integer range; the integer is what the Feistel cipher operates on. Ranges
# must not overlap and must fit in the 80-bit block (trivially so).
_PREFIX_GG = "TODO_GG_"      # 1..60
_PREFIX_GGX = "TODO_GGX_"    # 1..30 -> stored as 100..130
_GGX_OFFSET = 100


# ---------------------------------------------------------------------------
# 80-bit Feistel cipher (two 40-bit halves)
# ---------------------------------------------------------------------------

def _rotl40(x: int, n: int) -> int:
    """Rotate a 40-bit value left by ``n`` bits.

    Implemented as ``(x << n) | (x >> (40 - n))`` but with every shift count
    capped at <=13 so the pre-mask intermediate stays below 2**53 (the safe
    integer range on Flutter Web).
    """
    n %= _HALF_BITS
    if n == 0:
        return x & _HALF_MASK
    # Worst case here: x ~ 2**40 - 1, n = 13 -> x << n ~ 2**53 - 2**13 < 2**53.
    left = ((x << n) & _HALF_MASK)
    right = x >> (_HALF_BITS - n)
    return (left | right) & _HALF_MASK


def _f(half: int, k: int) -> int:
    """Round function: 40-bit -> 40-bit, keyed by ``k``.

    xorshift-style mixer using only XOR, add, shift and rotate so the Dart
    port mirrors it verbatim. Every shift count is <=13 so the worst-case
    pre-mask intermediate is ``(2**40 - 1) * 2**13 < 2**53`` — safe on
    Flutter Web.
    """
    x = (half + k) & _HALF_MASK
    x ^= _rotl40(x, 13)
    x = (x + ((x << 7) & _HALF_MASK)) & _HALF_MASK
    x ^= _rotl40(x, 11)
    x = (x + ((x << 5) & _HALF_MASK)) & _HALF_MASK
    x ^= _rotl40(x, 9)
    return x & _HALF_MASK


def _encrypt(left: int, right: int) -> tuple[int, int]:
    """Apply the forward Feistel network. Inputs and outputs are two 40-bit
    halves so the result fits in Dart's 64-bit ``int`` (the combined 80-bit
    value would not)."""
    left &= _HALF_MASK
    right &= _HALF_MASK
    for k in _MASTER_KEY:
        left, right = right, (left ^ _f(right, k)) & _HALF_MASK
    return left, right


def _decrypt(left: int, right: int) -> tuple[int, int]:
    """Invert :func:`_encrypt`."""
    left &= _HALF_MASK
    right &= _HALF_MASK
    for k in reversed(_MASTER_KEY):
        # Undo (L,R) -> (R, L XOR F(R,k))
        prev_right = left
        prev_left = (right ^ _f(left, k)) & _HALF_MASK
        left, right = prev_left, prev_right
    return left, right


# ---------------------------------------------------------------------------
# Base32 helpers
# ---------------------------------------------------------------------------

# Each 40-bit half encodes to exactly 8 base32 chars; the two halves are
# concatenated (high half first) to form the final 16-char code.
_HALF_CODE_LEN = _HALF_BITS // 5  # = 8


def _half_to_base32(n: int) -> str:
    """Encode a 40-bit unsigned int as exactly 8 base32 chars (big-endian)."""
    if n < 0 or n > _HALF_MASK:
        raise ValueError(f"value {n} does not fit in {_HALF_BITS} bits")
    chars: list[str] = []
    for _ in range(_HALF_CODE_LEN):
        chars.append(_ALPHABET[n & 31])
        n >>= 5
    return "".join(reversed(chars))


def _half_from_base32(code: str) -> int:
    """Decode an 8-char base32 chunk back into a 40-bit int. Case-insensitive."""
    n = 0
    for ch in code.upper():
        idx = _INDEX.get(ch)
        if idx is None:
            raise ValueError(f"invalid base32 character: {ch!r}")
        n = (n << 5) | idx
    return n & _HALF_MASK


# ---------------------------------------------------------------------------
# ID <-> integer mapping
# ---------------------------------------------------------------------------

def _id_to_int(card_id: str) -> int:
    """Map a known plaintext card ID to its integer slot."""
    if card_id.startswith(_PREFIX_GGX):
        suffix = card_id[len(_PREFIX_GGX):]
        idx = int(suffix)
        if not 1 <= idx <= 30:
            raise ValueError(f"GGX index out of range: {card_id}")
        return _GGX_OFFSET + idx
    if card_id.startswith(_PREFIX_GG):
        suffix = card_id[len(_PREFIX_GG):]
        idx = int(suffix)
        if not 1 <= idx <= 60:
            raise ValueError(f"GG index out of range: {card_id}")
        return idx
    raise ValueError(f"unknown card-ID shape: {card_id!r}")


def _int_to_id(n: int) -> str:
    """Inverse of :func:`_id_to_int`. Returns the canonical card-ID string."""
    if 1 <= n <= 60:
        return f"{_PREFIX_GG}{n:02d}"
    if _GGX_OFFSET + 1 <= n <= _GGX_OFFSET + 30:
        return f"{_PREFIX_GGX}{n - _GGX_OFFSET:02d}"
    raise ValueError(f"integer slot does not map to a known card: {n}")


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def encode_card_id(card_id: str) -> str:
    """Encode a plaintext card ID into its obfuscated 16-char code."""
    n = _id_to_int(card_id)
    # The plaintext lives in the low 40 bits; the high 40-bit half is zero.
    left, right = _encrypt((n >> _HALF_BITS) & _HALF_MASK, n & _HALF_MASK)
    return _half_to_base32(left) + _half_to_base32(right)


def decode_card_id(code: str) -> str:
    """Decode an obfuscated code back into its plaintext card ID.

    Raises ``ValueError`` if the code is malformed or does not correspond to
    a known card.
    """
    if len(code) != _CODE_LEN:
        raise ValueError(f"code must be {_CODE_LEN} chars, got {len(code)}")
    left = _half_from_base32(code[:_HALF_CODE_LEN])
    right = _half_from_base32(code[_HALF_CODE_LEN:])
    plain_left, plain_right = _decrypt(left, right)
    n = (plain_left << _HALF_BITS) | plain_right
    return _int_to_id(n)


def try_decode_card_id(code: str) -> str | None:
    """Lenient counterpart to :func:`decode_card_id`. Returns ``None`` instead
    of raising for invalid input."""
    try:
        return decode_card_id(code)
    except (ValueError, KeyError):
        return None


def all_known_card_ids() -> list[str]:
    """Return every plaintext card ID currently in the obfuscation domain.

    Useful for generators (Excel, QR HTML) that want to enumerate the whole
    catalogue without depending on the Dart mock data.
    """
    return (
        [f"{_PREFIX_GG}{i:02d}" for i in range(1, 61)]
        + [f"{_PREFIX_GGX}{i:02d}" for i in range(1, 31)]
    )


if __name__ == "__main__":  # pragma: no cover - manual smoke test
    for cid in all_known_card_ids():
        code = encode_card_id(cid)
        back = decode_card_id(code)
        assert back == cid, (cid, code, back)
        print(f"{cid:<14s} -> {code}")
